controllers/topology/desired_state.go

```diff
@@ -298,7 +382,96 @@ func computeMachineDeployment(_ context.Context, s *scope.Scope, machineDeployme
 	desiredMachineDeploymentObj.Spec.Replicas = machineDeploymentTopology.Replicas
 
 	desiredMachineDeployment.Object = desiredMachineDeploymentObj
-	return desiredMachineDeployment, nil
+	return desiredMachineDeployment, upgraded, nil
+}
+
+// computeMachineDeploymentVersion calculates the version of the desired machine deployment.
+// The version if calculated using the the state of the current machine deployments, the current control plane
+// and the version defined in the topology.
+func computeMachineDeploymentVersion(s *scope.Scope, desiredControlPlaneState *scope.ControlPlaneState, currentMDState *scope.MachineDeploymentState, allowUpgrade bool) (string, bool, error) {
+	desiredVersion := s.Blueprint.Topology.Version
+	// If creating a new machine deployment, we can pick up the desired version
+	// Note: We are not blocking the creation of new machine deployments when
+	// the control plane or any of the machine deployments are upgrading/scaling.
+	// TODO: Double check if it is okay to add a new machine deployment if the control plane is upgrading.
+	if currentMDState == nil || currentMDState.Object == nil {
+		return desiredVersion, false, nil
+	}
+
+	// get the current version of the machine deployment
+	currentVersion := *currentMDState.Object.Spec.Template.Spec.Version
+
+	// return early if we are not allowed to upgrade the machine deployment
+	if !allowUpgrade {
+		return currentVersion, false, nil
+	}
+
+	// return early if the currentVersion is already equal to the desiredVersion
+	// no further checks required.
+	if currentVersion == desiredVersion {
+		return currentVersion, false, nil
+	}
+
+	if s.Current.ControlPlane == nil || s.Current.ControlPlane.Object == nil {
+		// We dont have a control plane at this time but we have a machine deployment. (Is this possible? Manually deleted?).
+		// Do not perform any machine deployment upgrade in this case. Return the current version of the machine deployment.
+		return currentVersion, false, nil
+	}
```

Comment:
If I'm not wrong this should be 
```suggestion
	// If the control plane is being created (current control plane is nil), do not perform any machine deployment upgrade in this case. Return the current version of the machine deployment.
	// NOTE: this case should never happen (upgrading a MachineDeployment) before creating a CP, but we are implementing this check for extra safety.
	if s.Current.ControlPlane == nil || s.Current.ControlPlane.Object == nil {
		return currentVersion, false, nil
	}
```

Reply:
Yes, this is definitely a safety check. This should never happen in reality.